apps/prairielearn/elements/pl-drawing/elements.py

```diff
@@ -36,28 +39,35 @@ def abserr_ang(ref, x):
 
 
 class BaseElement:
-    def generate(element, data):
+    @staticmethod
+    def generate(el: HtmlElement, data: dict) -> dict:
         return {}
 
-    def is_gradable():
+    @staticmethod
+    def is_gradable() -> bool:
         return False
 
-    def grade(ref, student, tol, angtol):
+    @staticmethod
+    def grade(ref: dict, st: dict, tol: float, angtol: float) -> bool:
         return True
 
-    def grading_name(element):
+    @staticmethod
+    def grading_name(element: HtmlElement) -> str | None:
         return None
 
-    def validate_attributes():
+    @staticmethod
+    def validate_attributes() -> bool:
         return True
 
-    def get_attributes():
+    @staticmethod
+    def get_attributes() -> list[str]:
         """Return a list of attributes that the element may contain."""
         return []
 
 
 class ControlledLine(BaseElement):
-    def generate(el, data):
+    @staticmethod
+    def generate(el: HtmlElement, data: dict):
```

Comment:
No return type? Should be included explicitly even if it's just `dict`.

Reply:
What value does this add? Since the method is inherited, it will raise an error if the type isn't consistent with the `BaseElement` `generate` signature.

Reply:
If nothing else, makes it easier for someone who's reading just this method to understand what it's supposed to return without having to look at the base class. I also think that this won't raise an error if a type that inherits from a dict is returned?

My rule of thumb is to be as explicit as possible with Python types, since the typechecking can be pretty inconsistent, and makes the code easier to read in general.

Reply:
Changed in [3ffec2a](https://github.com/PrairieLearn/PrairieLearn/pull/11855/commits/3ffec2aa6208ba6b813e6d255452113ca0c8b8b9). While I agree generally, this code still has lots of type-errors and needs to be rewritten, so catching stuff like a type inherited from a dict doesn't really matter to me. According to my editor, there are 69 unresolved type errors still.

Reply:
That's not surprising, unfortunately the code for this element is a bit of a mess.